server/src/lib/server.rs

```diff
@@ -498,6 +454,41 @@ async fn instance_ensure(
             ))
         })?;
 
+    // Is this part of a migration?
+    let migrate = if let Some(migrate_request) = request.migrate {
+        // We stop short of the usual intialization routines because this is
+        // a migrate request and so we should try to establish a connection
+        // with the source instance.
+        let rqctx = rqctx.clone();
+        let res = migrate::dest_initiate(rqctx, instance_id, migrate_request)
+            .await
+            .map_err(<_ as Into<HttpError>>::into)?;
+        Some(res)
+    } else {
+        instance.on_transition(Box::new(move |next_state, _inv, ctx| {
+            match next_state {
+                propolis::instance::State::Boot => {
+                    // Set vCPUs to their proper boot (INIT) state
+                    for mut vcpu in ctx.mctx.vcpus() {
+                        vcpu.reboot_state().unwrap();
+                        vcpu.activate().unwrap();
+                        // Set BSP to start up
+                        if vcpu.is_bsp() {
+                            vcpu.set_run_state(bhyve_api::VRS_RUN).unwrap();
+                            vcpu.set_reg(
+                                bhyve_api::vm_reg_name::VM_REG_GUEST_RIP,
+                                0xfff0,
+                            )
+                            .unwrap();
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }));
+
+        None
+    };
     instance.print();
 
     // Save the newly created instance in the server's context.
```

Comment:
Need to move this above the `if let` and `drop(context)` before `dest_initiate` as it tries to lock the Mutex again leaving us deadlocked.

Reply:
Ooo, that's embarrassing. Done.